test/e2e/cluster_samples_operator_test.go

```diff
@@ -940,3 +963,92 @@ func TestRecreateDeletedManagedSample(t *testing.T) {
 	verifyDeletedTemplatesRecreated(t)
 	t.Logf("Config after TestRecreateDeletedManagedSample: %#v", verifyOperatorUp(t))
 }
+
+func TestUpgrade(t *testing.T) {
+	cfg := verifyOperatorUp(t)
+	err := verifyConditionsCompleteSamplesAdded(t)
+	if err != nil {
+		dumpPod(t)
+		t.Fatalf("samples not stable at start of upgrade test %#v", verifyOperatorUp(t))
+	}
+
+	newVersion := kapis.Now().String()
+	t.Logf("current version %s version for upgrade %s", cfg.Status.Version, newVersion)
+
+	// update env to trigger upgrade
+	depClient := kubeClient.AppsV1().Deployments("openshift-cluster-samples-operator")
+	err = wait.PollImmediate(1*time.Second, 1*time.Minute, func() (bool, error) {
+		dep, err := depClient.Get("cluster-samples-operator", metav1.GetOptions{})
+		if err != nil {
+			t.Logf("error waiting for operator deployment to exist: %v\n", err)
+			return false, nil
+		}
+		t.Logf("found operator deployment")
+		for i, env := range dep.Spec.Template.Spec.Containers[0].Env {
+			t.Logf("looking at env %s", env.Name)
+			if strings.TrimSpace(env.Name) == "RELEASE_VERSION" {
+				t.Log("updating RELEASE_VERSION env")
+				dep.Spec.Template.Spec.Containers[0].Env[i].Value = newVersion
+				_, err := depClient.Update(dep)
+				if err == nil {
+					return true, nil
+				}
+				t.Logf("%v", err)
+				if kerrors.IsConflict(err) {
+					return false, nil
+				}
+				if stub.IsRetryableAPIError(err) {
+					return false, nil
+				}
+				return false, err
+			}
+
+		}
+		return false, nil
+	})
+	if err != nil {
+		dumpPod(t)
+		t.Fatalf("problem updating deployment env")
+	}
+
+	err = wait.PollImmediate(1*time.Second, 1*time.Minute, func() (bool, error) {
+		cfg := verifyOperatorUp(t)
+		if cfg.ConditionTrue(samplesapi.MigrationInProgress) {
+			return true, nil
+		}
+		return false, nil
+	})
+
+	if err != nil {
+		dumpPod(t)
+		t.Fatalf("did not enter migration mode in time %#v", verifyOperatorUp(t))
+	}
+
+	err = wait.PollImmediate(1*time.Second, 3*time.Minute, func() (bool, error) {
+		cfg := verifyOperatorUp(t)
+		if cfg.Status.Version == newVersion {
+			return true, nil
+		}
+		return false, nil
+	})
+
+	verifyClusterOperatorConditionsComplete(t)
+
+	err = wait.PollImmediate(1*time.Second, 1*time.Minute, func() (bool, error) {
+		state, err := operatorClient.ClusterOperators().Get(operator.ClusterOperatorName, metav1.GetOptions{})
+		if err != nil {
+			t.Logf("%v", err)
+			return false, nil
+		}
+		if len(state.Status.Versions) > 0 && state.Status.Versions[0].Name == "operator" && state.Status.Versions[0].Version == newVersion {
+			return true, nil
+		}
+		t.Logf("CVO is %#v", state)
+		return false, nil
+	})
+	if err != nil {
+		t.Fatalf("cvo version not correct after upgrade %#v", verifyOperatorUp(t))
+	}
+
+	validateContent(t, nil)
```

Comment:
is this actually confirming the version annotation on the content somehow?  if not i'm missing where that is happening.

Reply:
it is ... the validation is a level or two down (and harder for you to see since I squashed the commit in this case)

I'll add review comments where those changes are